File: frontend/src/games/mental-rotation/core/levels.ts
/* psygames-mental-rotation-levels · VER 1 · 09.09.2026 */
/* LOCAL REV spatial-lab/2026-09-09.2 · psygames-codex-mac · not an app release */
import type {Axis} from './types';

export interface RotationLevelSpec {
  level:number; cubes:number; path:Axis[]; optionCount:number;
  foil:'mixed'|'one-cube'; change:string;
}

// Each row changes an actual task constraint. Human difficulty at transitions
// between cube counts still needs calibration; no claim of a psychometric scale.
const patterns:{path:Axis[];optionCount:number;foil:'mixed'|'one-cube';text:string}[]=[
  {path:['z'],optionCount:3,foil:'mixed',text:'90°/180° вокруг Z, три варианта'},
  {path:['z','z'],optionCount:3,foil:'mixed',text:'180°/270° вокруг Z, три варианта'},
  {path:['x','y'],optionCount:3,foil:'mixed',text:'X90° → Y90°/180°, три варианта'},
  {path:['x','y','y'],optionCount:4,foil:'mixed',text:'X90° → Y180°/270°, четыре варианта'},
  {path:['x','y','y'],optionCount:4,foil:'one-cube',text:'X90° → Y180°/270°; подделки отличаются перестановкой одного кубика'},
];
export const ROTATION_LEVELS:readonly RotationLevelSpec[]=Array.from({length:50},(_,i)=>{
  const p=patterns[i%5],cubes=4+Math.floor(i/5);
  return {level:i+1,cubes,path:[...p.path],optionCount:p.optionCount,foil:p.foil,change:`${cubes} кубиков: ${p.text}.`};
});
export function rotationLevelSpec(level:number):RotationLevelSpec {
  if(!Number.isFinite(level))throw new RangeError('rotation level');
  return ROTATION_LEVELS[Math.max(0,Math.min(49,Math.floor(level)-1))];
}

File: flutter/lib/games/mental_rotation/levels.dart
/// ЛЕСТНИЦА «ПОВОРОТА» — перенос `core/levels.ts` и `levelParams` из `core/rotation.ts`.
///
/// Пятьдесят ступеней: каждая пятёрка меняет путь поворота и подделки, каждая пятёрка выше —
/// добавляет кубик. Единственный источник и для экрана, и для генератора: подпись «4 кубика,
/// ось Z» на настройке обязана совпасть с тем, что выпало в партии.
library;

import 'geometry.dart';

class RotationLevelSpec {
  const RotationLevelSpec({
    required this.level,
    required this.cubes,
    required this.path,
    required this.optionCount,
    required this.foil,
  });

  final int level;
  final int cubes;
  final List<Axis> path;
  final int optionCount;

  /// `mixed` — зеркало и другая фигура; `one-cube` — подделки с переставленным кубиком.
  final String foil;
}

class _Pattern {
  const _Pattern(this.path, this.optionCount, this.foil);
  final List<Axis> path;
  final int optionCount;
  final String foil;
}

const List<_Pattern> _patterns = [
  _Pattern([Axis.z], 3, 'mixed'),
  _Pattern([Axis.z, Axis.z], 3, 'mixed'),
  _Pattern([Axis.x, Axis.y], 3, 'mixed'),
  _Pattern([Axis.x, Axis.y, Axis.y], 4, 'mixed'),
  _Pattern([Axis.x, Axis.y, Axis.y], 4, 'one-cube'),
];

final List<RotationLevelSpec> rotationLevels = List.generate(50, (i) {
  final p = _patterns[i % 5];
  return RotationLevelSpec(
    level: i + 1,
    cubes: 4 + (i ~/ 5),
    path: [...p.path],
    optionCount: p.optionCount,
    foil: p.foil,
  );
});

RotationLevelSpec rotationLevelSpec(int level) => rotationLevels[(level - 1).clamp(0, 49)];

/// Параметры пробы уровня: размер фигуры, оси подделок, число вариантов.
class LevelParams {
  const LevelParams({
    required this.minC,
    required this.maxC,
    required this.axes,
    required this.optionCount,
    required this.compound,
  });

  final int minC;
  final int maxC;

  /// Оси пути БЕЗ повторов — ими крутят подделки.
  final List<Axis> axes;
  final int optionCount;

  /// Составной поворот (косой ракурс) — с четвёртой ступени каждой пятёрки.
  final bool compound;
}

LevelParams levelParams(int level) {
  final s = rotationLevelSpec(level);
  final axes = <Axis>[];
  for (final a in s.path) {
    if (!axes.contains(a)) axes.add(a);
  }
  return LevelParams(
    minC: s.cubes,
    maxC: s.cubes,
    axes: axes,
    optionCount: s.optionCount,
    compound: axes.length > 1,
  );
}
